### DISCUSSION TITLE
What's a preferred way of dumping some debug information?

### DISCUSSION THREAD

**Discussion by storyteller-dev** [ctreffe/alfred#156] [Alfred Q&A] [answered]
Especially when learning a new framework, creating some temporary debug information can be beneficial. For instance, looking into an object I use in my script or simply putting out some variables when I don't know for certain what they actually hold.

However, `print()` does not seem to work or I overlooked something. The only reliable way of seeing such information I found was creating a Text element and putting it on some page. This has some limitations of course. I cannot see the data type of the variable or dump larger objects to see what they hold.

What would be a preferred way of dumping various types of information for debugging?

**Comment by jobrachem** [comment, marked as answer]
`print()` does work, but may be hard to spot in the console next to the log output. Here are my preferred methods in the order of preference:

1. Use the **debug mode of your IDE**. This is by far the most powerful debugging tool available. You can then set breakpoints instead of mere print statements. These breakpoints allow you to print out information from the current environment in an interactive console. You can also execute code line by line. It takes a little time to get used to a debugging environment, but in my view, it is one of the most valuable skills you can learn. Here's a tutorial for Pycharm: [link](https://www.jetbrains.com/help/pycharm/part-1-debugging-python-code.html#breakpoints).
2. Use **alfred3's logging system**. The `ExperimentSession`, and all `Page` and `Section` classes have `log` attribute, which gives access to a logging interface based on [Python's standard logging library](https://docs.python.org/3/howto/logging.html). You can use the methods `log.debug()`, `log.info()`, `log.warning()`, `log.error()`, `log.exception()`, and `log.log()`, which correspond to the methods offered by the standard library. *Be sure to check the log level in config.conf - it is set to 'info' by default, which means that debug log entries will be ignored.* The logging system is most useful when you are debugging alfred3 experiments that are hosted via mortimer, because you can then use mortimer's nice graphical display of log messages. In the console alone, log messages do look a little crowded.
3. Use a **text element**, as you described. It is virtually identical to using `print()` statements, but is easier to spot. As a drawbag, a text element will of course only be helpful if you have a working page that you can add it to.
4. Use a **print statement** and make it visible by including multiple newline characters `\n`, like `print("\n\n\nMytext\n\n\n")`. This will make the print statement much easier to spot. Print statements, like logging calls and the debug mode, have the advantage that they will never show up in an experiment like text elements will – which means that you cannot do much harm by accidentally forgetting to remove such a statement.

## Example for logging

```python
import alfred3 as al

exp = al.Experiment()

@exp.member
class Demo(al.Page):
    def on_exp_access(self):
        self.log.info("Info log message")
```
